### PR TITLE
CXX-1712 Overload to_json to correctly handle bson arrays

### PR DESCRIPTION
**Pull request by joshbsiegel**
- Overloaded to_json so that it handles `bsoncxx::array::view` and outputs the correct string
- Added test cases to adequately test the main four scenarios of an array being passed through to the function (legacy implicit, legacy explicit, relaxed, canonical)
- Tested locally in conjunction with https://github.com/mongodb/mongo-c-driver/pull/1305

### CODE REVIEW

**Review by kevinAlbs — COMMENTED** [review]
I suggest waiting until the C driver changes are merged before proceeding with this PR.

Once the C driver changes are merged, the C++ driver test and release scripts can be updated to use the unreleased commit of the C driver. I expect that requires changes in [.mci.yml here](https://github.com/mongodb/mongo-cxx-driver/blob/4d6685bf9c4e4acb0b853d0f0edcfcc77b09b0f3/.mci.yml#L10-L15) and [make_release.py here](https://github.com/mongodb/mongo-cxx-driver/blob/4d6685bf9c4e4acb0b853d0f0edcfcc77b09b0f3/etc/make_release.py#L86-L89)

File: src/mongocxx/CMakeLists.txt
Comment (kevinAlbs) [inline src/mongocxx/CMakeLists.txt:32]: Aside: I expect this is not necessary for this PR. `LIBMONGOC_REQUIRED_VERSION` controls the required version of libmongoc in cmake. It does not control the version of libmongoc obtained in testing. But this seems OK to include in this PR.

**Review by kevinAlbs — APPROVED** [review]

**Review by eramongodb — APPROVED** [review]
Minor suggestions/comments; otherwise, LGTM.

**Pull request merged**
